File: Leet 901/solution.py
from typing import NamedTuple


class Span(NamedTuple):
    price: int
    span: int


class StockSpanner:
    def __init__(self):
        self.stack: list[Span] = []

    def next(self, price: int) -> int:
        span: int = 1
        while self.stack and self.stack[-1].price <= price:
            span += self.stack.pop().span

        self.stack.append(Span(price, span))

        return span


def main() -> None:
    print("901. Online Stock Span")

    stock = StockSpanner()

    print(stock.next(100)) # return 1
    print(stock.next(80))  # return 1
    print(stock.next(60))  # return 1
    print(stock.next(70))  # return 2
    print(stock.next(60))  # return 1
    print(stock.next(75))  # return 4
    print(stock.next(85))  # return 6


if __name__ == "__main__":
    main()

File: Leet 901/solution.c
#include <stdio.h>
#include <stdlib.h>


typedef struct Span {
    int price;
    int amount;
} Span;


typedef struct Node {
    Span data;
    struct Node* next;
} Node;


typedef struct StockSpanner {
    Node* top;
} StockSpanner;


void stockPush(StockSpanner* self, Span value) {
    Node* node = malloc(sizeof(Node));
    node->data = value;
    node->next = self->top;
    self->top = node;
}


Span stockPop(StockSpanner* self) {
    Node* temp = self->top;
    Span value = temp->data;
    self->top = temp->next;
    
    free(temp);

    return value;
}


Span stockPeek(StockSpanner* self){ 
    return self->top->data;
}


int stockIsEmpty(StockSpanner* self) {
    return self->top == NULL;
}


void stockSpannerFree(StockSpanner* self) {
    while(!stockIsEmpty(self)) {
        stockPop(self);
    }

    free(self);
}


StockSpanner* stockSpannerCreate() {
    StockSpanner* stock = malloc(sizeof(StockSpanner));
    stock->top = NULL;
    return stock;
}


int stockSpannerNext(StockSpanner *self, int price) {
    int span = 1;
    while (!stockIsEmpty(self) && stockPeek(self).price <= price) {
        span += stockPop(self).amount;
    }

    stockPush(self, (Span){.price= price, .amount= span});
    return span;
}


int main() {
    StockSpanner *stock= stockSpannerCreate();

    printf("StockSpanner(100) = %d\n", stockSpannerNext(stock, 100));
    printf("StockSpanner(80) = %d\n", stockSpannerNext(stock, 80));
    printf("StockSpanner(60) = %d\n", stockSpannerNext(stock, 60));
    printf("StockSpanner(70) = %d\n", stockSpannerNext(stock, 70));
    printf("StockSpanner(60) = %d\n", stockSpannerNext(stock, 60));
    printf("StockSpanner(75) = %d\n", stockSpannerNext(stock, 75));
    printf("StockSpanner(85) = %d\n", stockSpannerNext(stock, 85));
    
    stockSpannerFree(stock);
    
    return 0;
}